File: Chanseok/프로그래머스/Lv2/17684. ［3차］ 압축/압축.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

class Solution {

    public int[] solution(String msg) {

        Map<String, Integer>  LZW = new HashMap<>();
        for(int i = 1; i <= 26; ++i) {
            LZW.put(String.valueOf((char)('A' + (i-1))), i);
        }

        ArrayList<Integer> answer = new ArrayList<>();

        String w, c;
        int len = msg.length();

	        int idx = 0;
	        while(idx < len) {
	        	w = c =  "";
	        	while(idx < len) {
	        		c = String.valueOf(msg.charAt(idx++));
	        		if(!LZW.containsKey(w+c)) {
	        			LZW.put(w+c, LZW.size()+1);
	        			c = "";
	        			break;
	        		}
	        		w += c;
	        	}
	        	
	        	if(c != "") {
	        		answer.add(LZW.get(w));
	        		break;
	        	}
	        	
	        	answer.add(LZW.get(w));
	        	--idx;
	        }

        int[] res = new int[answer.size()];
        for(int i = 0; i < answer.size(); ++i) res[i] = answer.get(i);
        return res;
    }
}

File: Seonguk/프로그래머스/압축/압축.py
def dfs(msg, idx, answer, dic):
    for i in range(idx+1, len(msg)+1):
        if msg[idx:i] not in dic: # 찾는 단어가 없으면 
            dic.append(msg[idx:i]) # 등록하고
            next_idx=i-1 # 다음 탐색 위치 저장
            break
        if i == len(msg):
            next_idx=i
            
    if next_idx == len(msg): # i가 끝에 도달했으면
        answer.append(dic.index(msg[idx:])) # 사전에 등록하고 종료
        return
    else:
        answer.append(dic.index(msg[idx:next_idx]))
    
    dfs(msg, next_idx, answer, dic)
    
def solution(msg):
    answer=[]
    dic=['','A','B','C','D','E','F','G','H','I','J','K','L','M','N',
         'O','P','Q','R','S','T','U','V','W','X','Y','Z']
    dfs(msg, 0, answer, dic)
    
    return answer
